From: xaizek <xaizek@openmailbox.org>
Date: Sun, 26 Oct 2014 23:25:30 +0200
Forwarded: yes
Last-Update: 2014-10-26
Description: fix building without terminal
 Fixes the colmgr-test to not fail when there is no terminal available.
 Big thanks to upstream for providing a fix on short notice!

---
 src/color_manager.c | 15 +++++++++------
 src/color_manager.h | 12 +++++++++++-
 src/vifm.c          | 10 +++++++++-
 tests/colmgr/test.c | 36 +++++++++++++++++++++++++++---------
 4 files changed, 56 insertions(+), 17 deletions(-)

diff --git a/src/color_manager.c b/src/color_manager.c
index 4e6f622..9a179f8 100644
--- a/src/color_manager.c
+++ b/src/color_manager.c
@@ -18,8 +18,6 @@
 
 #include "color_manager.h"
 
-#include <curses.h>
-
 #include <assert.h> /* assert() */
 #include <limits.h> /* CHAR_BIT */
 #include <stddef.h> /* NULL size_t */
@@ -44,10 +42,15 @@ static char *color_pair_map;
 /* Size of the color_pair_map in chars. */
 static size_t color_pair_map_size;
 
+static colmgr_conf_t conf;
+
 void
-colmgr_init(int max_color_pairs)
+colmgr_init(const colmgr_conf_t *conf_init)
 {
-	avail_pairs = max_color_pairs - FCOLOR_BASE;
+	assert(conf_init != NULL && "Init structure is required");
+	conf = *conf_init;
+
+	avail_pairs = conf.max_color_pairs - FCOLOR_BASE;
 	assert(avail_pairs >= 0 && "Too few color pairs available.");
 
 	color_pair_map_size = DIV_ROUND_UP(avail_pairs, CHAR_BIT);
@@ -102,7 +105,7 @@ static int
 color_pair_matches(int pair, int fg, int bg)
 {
 	short pair_fg, pair_bg;
-	pair_content(pair, &pair_fg, &pair_bg);
+	conf.pair_content(pair, &pair_fg, &pair_bg);
 	return (pair_fg == fg && pair_bg == bg);
 }
 
@@ -115,7 +118,7 @@ allocate_pair(int fg, int bg)
 	{
 		if(!get_bit(color_pair_map, i))
 		{
-			init_pair(FCOLOR_BASE + i, fg, bg);
+			conf.init_pair(FCOLOR_BASE + i, fg, bg);
 			set_bit(color_pair_map, i);
 			break;
 		}
diff --git a/src/color_manager.h b/src/color_manager.h
index c0b9bc0..8efb8af 100644
--- a/src/color_manager.h
+++ b/src/color_manager.h
@@ -19,10 +19,20 @@
 #ifndef VIFM__COLOR_MANAGER_H__
 #define VIFM__COLOR_MANAGER_H__
 
+typedef struct
+{
+	int max_color_pairs;
+	int (*init_pair)(short pair, short f, short b);
+	int (*pair_content)(short pair, short *f, short *b);
+}
+colmgr_conf_t;
+
 /* Initializes color manager unit.  Creates and prepares internal variables. */
-void colmgr_init(int max_color_pairs);
+void colmgr_init(const colmgr_conf_t *conf_init);
+
 /* Resets all color pairs that are available for dynamic allocation. */
 void colmgr_reset(void);
+
 /* Dynamically allocates color pair of specified foreground (fg) and background
  * (bg) colors.  Returns -1 on allocation failure. */
 int colmgr_alloc_pair(int fg, int bg);
diff --git a/src/vifm.c b/src/vifm.c
index f79d7a1..536ff7e 100644
--- a/src/vifm.c
+++ b/src/vifm.c
@@ -386,7 +386,15 @@ main(int argc, char *argv[])
 	if(!setup_ncurses_interface())
 		return -1;
 
-	colmgr_init(COLOR_PAIRS);
+	{
+		const colmgr_conf_t colmgr_conf = {
+			.max_color_pairs = COLOR_PAIRS,
+			.init_pair = &init_pair,
+			.pair_content = &pair_content,
+		};
+		colmgr_init(&colmgr_conf);
+	}
+
 	init_modes();
 	init_undo_list(&undo_perform_func, NULL, &ui_cancellation_requested,
 			&cfg.undo_levels);
diff --git a/tests/colmgr/test.c b/tests/colmgr/test.c
index db28ca2..08cceb2 100644
--- a/tests/colmgr/test.c
+++ b/tests/colmgr/test.c
@@ -1,31 +1,49 @@
 #include "test.h"
 
-#include <curses.h>
-
-#include <assert.h>
-
 #include "seatest.h"
 
+#include "../../src/utils/macros.h"
 #include "../../src/color_manager.h"
 
 void basic_tests(void);
 
+static int colors[64][2];
+
 static void
 all_tests(void)
 {
 	basic_tests();
 }
 
+static int
+init_pair(short pair, short f, short b)
+{
+	colors[pair][0] = f;
+	colors[pair][1] = b;
+	return 0;
+}
+
+static int
+pair_content(short pair, short *f, short *b)
+{
+	*f = colors[pair][0];
+	*b = colors[pair][1];
+	return 0;
+}
+
 int
 main(void)
 {
 	int result;
 
-	initscr();
-	assert(has_colors() && "Terminal should support colors.");
-	start_color();
-	colmgr_init(MAX_COLOR_PAIRS);
-	endwin();
+	{
+		const colmgr_conf_t colmgr_conf = {
+			.max_color_pairs = ARRAY_LEN(colors),
+			.init_pair = &init_pair,
+			.pair_content = &pair_content,
+		};
+		colmgr_init(&colmgr_conf);
+	}
 
 	result = run_tests(all_tests);
 
